### ISSUE TITLE
Is it possible to detect missing packets and packet drop?

### ISSUE THREAD

**Issue by luckygg** [seladb/PcapPlusPlus#954]
Hi all.

I have one simple question.

I would like to detect missing packets and packet drops.
Is it possible to check using PcapPlusPlus?
If possible, could you please how to check missing packets and drops?

Thank you.

**Comment by seladb** [comment]
Yes, this is possible. Please take a look at [`PcapLiveDevice::getStatistics()`](https://github.com/seladb/PcapPlusPlus/blob/22269e8d40bebdbef225d63bf0a92a9bd94988d7/Pcap%2B%2B/header/PcapLiveDevice.h#L548). 

You can also find information in the API reference: https://pcapplusplus.github.io/api-docs/next/classpcpp_1_1_pcap_live_device.html#a63c26fadf6e36c99608f500f67d60f29

**Comment by luckygg** [comment]
@seladb Thank you for your reply!
But I have still some questions.

I think [PcapLiveDevice::getStatistics()](https://github.com/seladb/PcapPlusPlus/blob/22269e8d40bebdbef225d63bf0a92a9bd94988d7/Pcap%2B%2B/header/PcapLiveDevice.h#L548) is current device's status.

void packetArrives(pcpp::RawPacket* packet)
{
	pcpp::Packet parsedPacket(packet);
        //In this code, how to check 'packet' has missing packets or drops?
}

**Comment by seladb** [comment]
If a packet was dropped the `packetArrives()` callback won't be called...

Oh wait, let me explain: as you mentioned, [`PcapLiveDevice::getStatistics()`](https://github.com/seladb/PcapPlusPlus/blob/22269e8d40bebdbef225d63bf0a92a9bd94988d7/Pcap%2B%2B/header/PcapLiveDevice.h#L548) reflects the device status, meaning how many packets were received by the NIC but could not be captured which can happen from a variety of reasons, for example traffic overload or slow packet processing. If a packet was dropped it won't be processed by PcapPlusPlues, hence the callback won't be called.

Maybe the missing or dropped packets you're referring to are something else like packets dropped by the OS? 🤔

**Comment by luckygg** [comment]
@seladb  Thank you for your kind explanation! :)

Actually, I have lack of network knowledge, so my question would be wrong.
I don't know where the missing packet would be occurred.
I was just worried and wondered if it could be detected.

I have an additional question!
The below image is the WireSharks's screenshot.
What function can I get the part marked in red from in PcapPlusPlus?
![wireshark](https://user-images.githubusercontent.com/17895255/188306947-2d494830-59bc-491c-8fa1-e9a31dff428b.png)

I'm using the code as like below...
void CustomWireshark::packetArrives(pcpp::RawPacket* packet)
{
	pcpp::Packet parsedPacket(packet);
	packet->getFrameLength(); //Wireshark's Frame Length?
	packet->getRawDataLen(); //Wireshark's Captured Length?
        //How to get the frame number?!
}

Thank you :)

**Comment by seladb** [comment]
I'm sorry for the delayed response... you are correct:

```cpp
void CustomWireshark::packetArrives(pcpp::RawPacket* packet)
{
    // This is 2 and 5
    packet->getRawDataLen();
    // This is 1 and 4
    packet->getFrameLength();
}
```

Please notice that when capturing from a live device 1 == 2 == 4 == 5 and all of them represent the captured length. Frame length is not available on live devices.

When reading from pcap/pcapng files there is (or might be) a difference between 2,5 and 1,4.

Regarding 3 - this is just a running counter of the captured packets, hence not available in PcapPlusPlus, but you can easily implement it yourself.

Please let me know if this helps.

**Comment by luckygg** [comment]
@seladb Thank you for your detailed explanation!
All my questions have been resolved.

Thank you :)
